[PM-3883] Fix ConfigService.serverConfig$ initial values and error handling

eliykat opened the pull request:
## Type of change

<!-- (mark with an `X`) -->

```
- [x] Bug fix
- [ ] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

<!--Describe what the purpose of this PR is. For example: what bug you're fixing or what new feature you're adding-->

**This will be picked to rc**

Per #5602, `configService.serverConfig$` emits values in the following order after initialization:

1. the stored `serverConfig` (or `null` if there is none)
2. then the latest `serverConfig` from server (if possible)

There are several problems with this:
* callers using `firstValueFrom` on this observable will usually get an out of date/null value, because they'll get the first value but not the second. This is a limitation of using `firstValueFrom`, but it does seem odd that we don't try to emit the latest-and-greatest as the first preference
* this is particularly a problem with CLI, which executes as a one-shot and can't hang around for updated values. (But even subscribers in the other apps would quickly flicker through several states)
* there is no error handling for the fetch from server, and if an observable errors out then no further values are emitted, so the whole thing falls over

This PR fixes this by _always_ getting the latest-and-greatest value from server and emitting that as the first value, and falling back to the stored value only if that is not possible. Note that this does not result in any extra server calls; if anything, it saves an unnecessary read from disk.

Incidentally, this also means that the stream recovers from a network error.

We could use the value from disk _instead_ of fetching from server if it's sufficiently recent and matches the current environmentUrls, but that's out of scope here. I'm just trying to get the current implementation working.

## Code changes

<!--Explain the changes you've made to each file or major component. This should help the reviewer understand your changes-->
<!--Also refer to any related changes or PRs in other repositories-->

- **file.ext:** Description of what was changed and why

## Screenshots

<!--Required for any UI changes. Delete if not applicable-->

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team
- Ensure that all UI additions follow [WCAG AA requirements](https://contributing.bitwarden.com/contributing/accessibility/)

eliykat commented on libs/common/src/platform/services/config/config.service.ts:
```diff
@@ -48,27 +50,22 @@ export class ConfigService implements ConfigServiceAbstraction {
         map((response) => new ServerConfigData(response)),
         delayWhen((data) => this.saveConfig(data)),
-        map((data) => new ServerConfig(data))
+        catchError((e: unknown) => {
+          // fall back to stored ServerConfig (if any)
+          this.logService.error("Unable to fetch ServerConfig: " + (e as Error)?.message);
```
I personally expect this to always be an `Error` object, but the rxjs documentation says it's an `any` type, so this seemed like the most correct way to represent its type.

Hinton replied:
It's our code that throws, so they are correct that it doesn't have to be an `Error`.

coroiu replied:
Could also just write ```this.logService.error(`Unable to fetch ServerConfig: ${e}`);``` and let js convert to appropriately string

eliykat commented:
Turns out this does not properly handle the error inside the inner observable created by `concatMap`. I fell back to moving more logic back inside a promise (RIP my nice rxjs data flow) because the rxjs solution was excessively complicated.

eliykat commented:
Turns out that the previous solution didn't work, because `catchError` did not save the inner observable from erroring out. `catchError` needs to be applied to the inner observable in `concatMap`. I've added a test, fixed it and also ran it past @coroiu.

Hinton commented on libs/common/src/platform/services/config/config.service.ts:
```diff
@@ -48,27 +50,25 @@ export class ConfigService implements ConfigServiceAbstraction {
-    // Get config from storage on initial load
-    const fromStorage = from(this.stateService.getServerConfig()).pipe(
-      map((data) => (data == null ? null : new ServerConfig(data)))
+    const latestServerConfig$ = defer(() => this.configApiService.get()).pipe(
```
Is `defer` needed?

eliykat replied:
Yes - my understanding is that we're declaring it outside the `concatMap` so we need to convert the promise into an observable explicitly, either `from` or `defer`. `from` gives us a one-shot execution, `defer` creates a factory for each subscriber, which is what we want here so that the server fetch can be repeated over time.
